cpp/ql/src/semmle/code/cpp/MemberFunction.qll

```diff
@@ -197,8 +197,12 @@ class Constructor extends MemberFunction, TaintFunction {
 
   override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
     // taint flow from any constructor argument to the returned object
-    input.isParameter(_) and
-    output.isReturnValue()
+    exists(int idx |
+      input.isParameter(idx) and
+      output.isReturnValue() and
+      not this.(CopyConstructor).hasDataFlow(input, output) and // don't duplicate where we have data flow
+      not this.(MoveConstructor).hasDataFlow(input, output) // don't duplicate where we have data flow
```

Comment:
Can these be merged into one as `this.(DataFlowFunction).hasDataFlow(input, output)`?

Reply:
Yes, this works, good suggestion ... but it turns out this bits getting deleted anyway.